### ISSUE TITLE
Add mbtiles binary diffing and patching (per tile)

### ISSUE THREAD

**Issue by nyurik** [maplibre/martin#1049]
TLDR; reduce the size of a delta tileset, storing binary diffs instead of full tiles.

## What we have now
Currently, `mbtiles` offers a way to compare two tilesets A and B and create a new "patch" tileset, which only contains tiles that changed between A and B.  The patch tileset behaves just like the real mbtiles file, except that it also contains `NULL` tiles to indicate that a tile in A does not exist in B, and should be deleted.

## Proposal
Introduce a new type of a "patch" that supports binary diffs between individual tiles in a tileset.  We could use the following architecture.

* Depending on the type of an mbtiles file (`flat` / `flat-with-hash` / `normalized`), create a matching "binary diff" tables.
* For `flat` and `flat-with-hash`, add a new table `tiles_bsdiff` with columns:
    `zoom_level, tile_column, tile_row, bsdiff (blob), old_hash (text), new_hash (text)`
  * bsdiff is a value generated by comparing two tiles with a library like [qbsdiff](https://github.com/hucsmn/qbsdiff), [bsdiff-rs](https://github.com/space-wizards/bsdiff-rs), or some other one.  Binary diffing standard is preferred, but so is performance.
  * `old_hash` and `new_hash` would use the same algorithm for hashing as the rest of the patch file (currently MD5)
  * When applying with `apply-patch`, it will validate that the tile at the `z/x/y` location has the same hash as `old_hash`, apply the diff, and verify that it matches the new hash.
  * When generating a delta, the differ should validate that applying generated diff to the original tile produces desired tile.  This will ensure that we can rely on the patch result.
* For `normalized` structure, the table would be `images_bsdiff` with columns:
    `old_id, new_id, bsdiff`
  * The patch will be applied to the `images` table, only adding new rows (cleanup of the unused tiles will happen after all other file manipulations are done)
  * New rows will have `new_id` as their `tile_id` value
  * Both `old_id` and `new_id` are expected to be the hashes of the old/new tile value
  * It is possible for a patch to contain multiple identical `old_id` (but different `new_id`). E.g. a tile with value X was modified to become X1 and X2 (for different x/y/z)
* The new binary diff table will exist in addition to the regular mbtiles structure
  * It is possible for the diffing algorithm to store some tiles as a regular replacement (current functionality), and some tiles as binary diffs (in the new table). Only one method is allowed per tile - a tile could either be binary-diffed or replaced, but not both.
* When applying, first apply the binary changes (this way normalized `images` table gets the new values), followed by the regular apply-diff process.

## TBD
* Is it worth trying to preserve the original mbtiles format and expectations, or should this be some `.mbtiles-diff` file type?
  * The current diff format "technically" does not break the mbtiles spec, with the possible exception of `NULL` values being used as a tile (which seems OK from the spec's perspective)
  * The new bsdiff format may break expectations because the `normalized` *map* table may contain `z/x/y/id`, but the `images` table will not have that `id`. So joining them in the `tiles` view would either skip that row, or if defined as `LEFT JOIN`, would give `tile_data` as `NULL`. Neither of these cases technically break the spec, but may result in an appearance of an empty mbtiles file.
* What mechanism should be used to ensure compatibility?
  * Is it ok to apply diff to a file X? Maybe store a hash value of the old file?

**Comment by CommanderStorm** [comment]
- https://github.com/maplibre/martin/pull/1358
